src/Server.cc

```diff
@@ -148,31 +152,127 @@ Server::Server(const ServerConfig &_config)
+    }
+
     // \todo(nkoenig) Async resource download.
     // This call can block for a long period of time while
     // resources are downloaded. Blocking here causes the GUI to block with
     // a black screen (search for "Async resource download" in
     // 'src/gui_main.cc'.
-    errors = this->dataPtr->sdfRoot.Load(filePath);
+    errorsThread = root.Load(filePath);
+    this->dataPtr->sdfRoot = root;
+
+    if (!this->dataPtr->downloadInParallel)
+      return true;
+
+    if (this->dataPtr->sdfRoot.WorldCount() == 0)
+    {
+      ignerr << "There is no world available" << "\n";
+      return false;
+    }
+
+    while (this->dataPtr->simRunners.size() == 0)
+    {
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+
+    for (std::unique_ptr<SimulationRunner> &runner :
+         this->dataPtr->simRunners)
+    {
+      for (size_t j = 0; j < this->dataPtr->sdfRoot.WorldCount(); ++j)
+      {
+          runner->AddWorld(this->dataPtr->sdfRoot.WorldByIndex(j));
+      }
+      runner->SetFetchedAllIncludes(true);
+    }
+    ignmsg << "Download models in parallel has finished. "
+           << "Now you can start the simulation" << std::endl;
   }
   else
   {
     ignmsg << "Loading default world.\n";
+
+    // Load an empty world.
+    /// \todo(nkoenig) Add a "AddWorld" function to sdf::Root.
+    errorsThread = root.LoadSdfString(DefaultWorld::World());
+    this->dataPtr->sdfRoot = root;
+  }
+
+  if (!errorsThread.empty())
+  {
+    for (auto &err : errorsThread)
+      ignerr << err << "\n";
+    return false;
+  }
+
+  return true;
+}
+
+/////////////////////////////////////////////////
+void Server::Init()
+{
+  // Configure the fuel client
+  fuel_tools::ClientConfig config;
+  if (!this->dataPtr->config.ResourceCache().empty())
+    config.SetCacheLocation(this->dataPtr->config.ResourceCache());
+  this->dataPtr->fuelClient = std::make_unique<fuel_tools::FuelClient>(config);
+
+  // Configure SDF to fetch assets from ignition fuel.
+  sdf::setFindCallback(std::bind(&ServerPrivate::FetchResource,
+        this->dataPtr.get(), std::placeholders::_1));
+  common::addFindFileURICallback(std::bind(&ServerPrivate::FetchResourceUri,
+      this->dataPtr.get(), std::placeholders::_1));
+
+  addResourcePaths();
+
+  sdf::Errors errors;
+
+  if (this->dataPtr->downloadInParallel)
+  {
+    this->dataPtr->downloadModelsThread = std::thread([&]()
+    {
+      this->DownloadModels();
+    });
+
+    ignmsg << "Loading default world.\n";
+
+    common::SystemPaths systemPaths;
+    // Worlds from environment variable
+    systemPaths.SetFilePathEnv(kResourcePathEnv);
+    // Worlds installed with ign-gazebo
+    systemPaths.AddFilePaths(IGN_GAZEBO_WORLD_INSTALL_DIR);
+    std::string filePath = systemPaths.FindFile(this->dataPtr->config.SdfFile());
+
+    std::string worldName;
+    auto errors2 = this->dataPtr->sdfRoot.GetWorldName(filePath, worldName);
```

Comment:
This is giving me compilation errors. I see https://github.com/gazebosim/sdformat/pull/1027 targeting Garden, it doesn't look like there are breaking changes. Should it target `sdf9` so that it works with this PR? Or are you planning on closing this PR in favor of https://github.com/gazebosim/gz-sim/pull/1484?

Reply:
you need this branch https://github.com/gazebosim/sdformat/tree/ahcorde/sdf9/worldName

Both PR are doing the same thing, I want to include this on citadel, but I have some issues here that are not happing in Garden.